src/python/grpcio_tests/tests/unit/_metadata_flags_test.py

```diff
@@ -0,0 +1,244 @@
+        timeout=test_constants.LONG_TIMEOUT,
+        wait_for_ready=wait_for_ready)
+
+
+def perform_stream_unary_future(channel, wait_for_ready=None):
+    channel.stream_unary(_STREAM_UNARY).future(
+        iter([_REQUEST] * test_constants.STREAM_LENGTH),
+        timeout=test_constants.LONG_TIMEOUT,
+        wait_for_ready=wait_for_ready).result(
+            timeout=test_constants.LONG_TIMEOUT)
+
+
+def perform_stream_stream_call(channel, wait_for_ready=None):
+    response_iterator = channel.stream_stream(_STREAM_STREAM).__call__(
+        iter([_REQUEST] * test_constants.STREAM_LENGTH),
+        timeout=test_constants.LONG_TIMEOUT,
+        wait_for_ready=wait_for_ready)
+    for _ in response_iterator:
+        pass
+
+
+_ALL_CALL_CASES = [
+    perform_unary_unary_call, perform_unary_unary_with_call,
+    perform_unary_unary_future, perform_unary_stream_call,
+    perform_stream_unary_call, perform_stream_unary_with_call,
+    perform_stream_unary_future, perform_stream_stream_call
+]
+
+
+class MetadataFlagsTest(unittest.TestCase):
+
+    def check_connection_does_failfast(self, fn, channel, wait_for_ready=None):
+        try:
+            fn(channel, wait_for_ready)
+        except BaseException as e:  # pylint: disable=broad-except
+            self.assertIn('StatusCode.UNAVAILABLE', str(e))
+
+    def test_call_wait_for_ready_default(self):
+        for perform_call in _ALL_CALL_CASES:
+            self.check_connection_does_failfast(perform_call,
+                                                create_dummy_channel())
+
+    def test_call_wait_for_ready_disabled(self):
+        for perform_call in _ALL_CALL_CASES:
+            self.check_connection_does_failfast(
+                perform_call, create_dummy_channel(), wait_for_ready=False)
+
+    def test_call_wait_for_ready_enabled(self):
+        # To test the wait mechanism, Python thread is required to make
+        #   client set up first without handling them case by case.
+        # Also, Python thread don't pass the unhandled exceptions to
+        #   main thread. So, it need another method to store the
+        #   exceptions and raise them again in main thread.
+        self.unhandled_exceptions = queue.Queue()
+        tcp, addr = get_free_loopback_tcp_port()
+
+        def test_call(perform_call=None):
+            try:
+                channel = grpc.insecure_channel(addr)
+                perform_call(channel, wait_for_ready=True)
+            except BaseException as e:  # pylint: disable=broad-except
+                self.unhandled_exceptions.put(e, True)
+
+        test_threads = []
+        for perform_call in _ALL_CALL_CASES:
+            test_thread = threading.Thread(
+                target=test_call, args=(perform_call,))
+            test_thread.exception = None
+            test_thread.start()
+            test_threads.append(test_thread)
+
+        # Start the server after the connections are waiting
+        time.sleep(2)
```

Comment:
This seems like it could result in a flaky test. Is there a more intelligent mechanism we can use to determine when the connections are waiting?

Reply:
This one is a little bit awkward. If the program doesn't sleep here, even wait-for-ready set to `False` can pass the test. So I choose a sleep time larger then the retries, which is half of the  timeout. I think this is the balanced point to avoid flaky test.